test-apps/display-test-app/ios/imodeljs-test-app/imodeljs-test-app/DtaServiceAuthorizationClient.swift

```diff
@@ -0,0 +1,129 @@
+        if let authorityURL = URL(string: authority) {
+            if var authorityURLComponents = URLComponents(url: authorityURL, resolvingAgainstBaseURL: false) {
+                if prefix != nil {
+                    prefix = prefix == "dev-" ? "qa-" : prefix
+                    authorityURLComponents.host = prefix! + authorityURLComponents.host!
+                    if authorityURLComponents.path.hasSuffix("/") {
+                        authorityURLComponents.path.removeLast()
+                    }
+                    if let url = authorityURLComponents.url {
+                        self.issuerURL = url
+                    }
+                }
+            }
+        }
+    }
+
+    func checkSettings() -> NSError? {
+        if authSettings.clientId.count == 0 {
+            return error(reason: "DtaOidcAuthorizationClient: initialize() was called with invalid or empty clientId")
+        }
+        if authSettings.clientSecret.count == 0 {
+            return error(reason: "DtaOidcAuthorizationClient: initialize() was called with invalid or empty clientSecret")
+        }
+        if authSettings.scope.count == 0 {
+            return error(reason: "DtaOidcAuthorizationClient: initialize() was called with invalid or empty scope")
+        }
+        let scope = authSettings.scope
+        if scope.contains("openid") || scope.contains("email") || scope.contains("profile") || scope.contains("organization") {
+            return error(reason: "DtaOidcAuthorizationClient: Scopes for a service cannot include 'openid email profile organization'")
+        }
+        return nil
+    }
+
+    func generateAccessToken(_ completion: @escaping DtaServiceAuthorizationClientCallback) {
+        let session = URLSession(configuration: URLSessionConfiguration.default)
+        let requestURL = issuerURL.appendingPathComponent("connect").appendingPathComponent("token")
+        var request = URLRequest(url: requestURL)
+        request.setValue("application/json", forHTTPHeaderField: "Accept")
+        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
+        request.httpMethod = "POST"
+        let bodyString = "grant_type=client_credentials&scope=\(authSettings.scope.encodedForURLQuery()!)"
+        let bodyData = Data(bodyString.utf8)
+        request.setValue("\(bodyData.count)", forHTTPHeaderField: "Content-Length")
+        request.httpBody = bodyData
+        let encoded = "\(authSettings.clientId.encodedForURLQuery()!):\(authSettings.clientSecret.encodedForURLQuery()!)"
+        request.setValue("Basic \(encoded.toBase64())", forHTTPHeaderField: "Authorization")
+        let task = session.dataTask(with: request) { (data, response, error) in
+            if error == nil {
+                if let response = response as? HTTPURLResponse
+                {
+                    if response.statusCode == 200 {
+                        if let data = data {
```

Comment:
Combine as one if statement.

Reply:
There are two different errors for the two different if statements. Combining them would mess that up.

Reply:
What I meant above is that the two have different else clauses.

Reply:
My line highlighting got messed up. These two lines can be combined:
```swift
                        if let data = data {
                            if let json = try? JSONSerialization.jsonObject(with: data) as? JSON {
```

Reply:
Ah, I noticed that myself after looking at the else clauses, and fixed it (as well as getting rid of a bug where the completion handler wasn't called).